### PR TITLE
Locale/utf8 fixes

### PR DESCRIPTION
**Pull request by techee**
I finally managed to test utf8/locale conversions in Geany and realised there are several problems these patches try to solve. No guarantee this fixes every single problem but at least the basic things work after this change.

### CODE REVIEW

**Comment by elextr** [comment]
Great work, I havn't looked at the changes in great detail yet, but I have long suspected that there were some utf8/locale confusions.

File: plugins/filebrowser.c [hunk 1]
```diff
@@ -233,7 +233,7 @@ static void add_item(const gchar *name)
 	sep = (utils_str_equal(current_dir, "/")) ? "" : G_DIR_SEPARATOR_S;
 	fname = g_strconcat(current_dir, sep, name, NULL);
 	dir = g_file_test(fname, G_FILE_TEST_IS_DIR);
-	utf8_fullname = utils_get_locale_from_utf8(fname);
+	utf8_fullname = utils_get_utf8_from_locale(fname);
```
Comment (b4n) [inline plugins/filebrowser.c]: haha nice catch!

File: src/build.c [hunk 2]
```diff
@@ -678,20 +677,18 @@ static gchar *build_replace_placeholder(const GeanyDocument *doc, const gchar *s
 	stack = g_string_new(src);
 	if (doc != NULL && doc->file_name != NULL)
 	{
-		filename = utils_get_utf8_from_locale(doc->file_name);
```
Comment (b4n) [inline src/build.c]: okay, wrong indeed as `doc->file_name` is already UTF-8

File: src/build.c [hunk 3]
```diff
@@ -712,17 +709,16 @@ static gchar *build_replace_placeholder(const GeanyDocument *doc, const gchar *s
...
 
 	utils_string_replace_all(stack, "%p", replacement);
 	g_free(replacement);
 
-	ret_str = utils_get_utf8_from_locale(stack->str);
+	ret_str = stack->str;
 	g_free(executable);
-	g_free(filename);
-	g_string_free(stack, TRUE);
+	g_string_free(stack, FALSE);
```
Comment (b4n) [inline src/build.c]: This should probably rather be rewritten as `return g_string_free(stack, FALSE);`

File: src/build.c [hunk 4]
```diff
@@ -750,15 +746,16 @@ static void build_spawn_cmd(GeanyDocument *doc, const gchar *cmd, const gchar *d
...
 
 	gtk_list_store_clear(msgwindow.store_compiler);
 	gtk_notebook_set_current_page(GTK_NOTEBOOK(msgwindow.notebook), MSG_COMPILER);
-	msgwin_compiler_add(COLOR_BLUE, _("%s (in directory: %s)"), utf8_cmd_string, utf8_working_dir);
+	msgwin_compiler_add(COLOR_BLUE, _("%s (in directory: %s)"), cmd, utf8_working_dir);
 	g_free(utf8_working_dir);
-	g_free(utf8_cmd_string);
+
+	locale_cmd_string = utils_get_locale_from_utf8(cmd);
+	argv[2] = locale_cmd_string;
```
Comment (b4n) [inline src/build.c]: okay, though apparently the naming scheme in this function is "utf8_" for UTF-8 and nothing for locale -- but probably the contrary would be better, so I'm fine with this.

**Comment by techee** [comment]
As suggested I did the "return g_string_free()" thing and removed the locale_ prefix in the function where utf8_ was used.

File: src/utils.c [hunk 5]
```diff
@@ -556,7 +556,7 @@ gchar *utils_str_middle_truncate(const gchar *string, guint truncate_length)
 
 	length = strlen(string);
 
-	g_return_val_if_fail(g_utf8_validate(string, length, NULL), NULL);
+	g_return_val_if_fail(g_utf8_validate(string, length, NULL), g_strdup(string));
```
Comment (eht16) [inline src/utils.c]: Wouldn't the `g_strdup(string)` leaked in case g_return_val_if_fail() does not return on fail?

Comment (techee) [inline src/utils.c]: Of course! Thanks for noticing, will fix.

**Comment by techee** [comment]
I just simply dropped the utils_str_middle_truncate() patch - I think a better way would be to use things like g_filename_to_utf8() where appropriate - I may have a look into this in the future.

**Comment by eht16** [comment]
Reviewed the diff and it looks fine however I did not test it in detail. Or as Lex would say: LGBI :)

**Comment by techee** [comment]
@b4n Would you consider this one for 1.27? I think it would be nice to have those utf8/locale fixed and I believe it isn't that complicated/risky to be applied now.

**Pull request merged**